### DISCUSSION TITLE
M2M deep permissions Create

### DISCUSSION THREAD

**Discussion by gkielwasser** [directus/directus#8589] [Archive]
Hi, 
I have the following collection:
```
Wishes_users: {
	products: M2M Products,
	user: M2O Directus_users
}
```

I'm trying to restrict using permissions the creation of new items in Wishes_users_products so that it must belongs to the user.

```
POST items/wishes_users_products
{
	"products_id" : "49c67628-eb93-4dc0-89d1-138d249b5c98",
	"wishes_users_id":"c64da757-de9a-4105-a897-37cad75a7ccc"
}
```
Permissions on Wishes_products (Create)
```
{
    "wishes_users_id": {
        "user": {
            "_eq": "$CURRENT_USER"
        }
    }
}
```

As soon as I have this permission defined, I have the following error: 
```
{
  "errors": [
    {
      "message": "\"wishes_users_id\" must be of type object",
      "extensions": {
        "code": "FAILED_VALIDATION",
        "field": "wishes_users_id"
      }
    }
  ]
}
```

Note: about read access of this collection, I have defined the same permission and it works well according to the need describe upper.

What is wrong with such a permission in the create ?

Thanks

**Comment by joselcvarela** [comment]
Hello @gkielwasser 
Can I ask why are you adding this permission?
Every time a user creates a wish, that wish should be assigned to the user, so no permissions needed. Is that correct?

Also, I think it's not possible to do this kind of validation because the validation happens before making any change into database and since the record is not present yet on the database, it will always fail.

**Reply by gkielwasser** [reply]
Hi @joselcvarela ,

given this model I could add wish by 2 ways.

First way:
```
PATCH items/wishes_users/a8438c10-5ce0-42ae-9e7d-6da48d73525c
{
	"products" : [
		{"products_id":"6af60f90-c097-4447-8f56-c27b88643483"},
		{"products_id":"49c67628-eb93-4dc0-89d1-138d249b5c98"}
	]
}
```
Cons: in order to add/delete a wish we have to "rebuild" the products array each time. 
It implies to have the exact previous array before to performs the mutation.
For instance, to add an item we have to "know" the previous entries to not override them. 

Seconday way:
```
POST items/wishes_users_products
{
	"products_id" : "49c67628-eb93-4dc0-89d1-138d249b5c98",
	"wishes_users_id":"c64da757-de9a-4105-a897-37cad75a7ccc"
}
DELETE items/wishes_users_products/:id
```
Pros: we don't have to know and manage the products array, we just tell what we want to add/remove.
Worst case that could happen in case of concurrency or whatever would be that the request would fail.


Happy to have feedback about that, I have to say, I have not choosed one or the other as for now :)

Anyway, for both method I must add on the collection Wishes_users_products a create permission on at least those fields ['wishes_users_id','products_id'], otherwise I do not have enough permission to add item.
Doing it, I open the create api of this collection for anyone. That is the reason why I was trying to restrict it.

**Reply by azrikahar** [reply]
Regarding the cons for the first way, you can add/delete by using the detailed syntax such as:

```js
// PATCH items/whishes_users/a8438c10-5ce0-42ae-9e7d-6da48d73525c
{
	"products" : [
		{ "name": "create new product as an example" }
	]
}
```
Link to this in the docs (need to scroll slightly down to "Detailed"): https://docs.directus.io/reference/api/introduction/#one-to-many-many-to-many

**Reply by gkielwasser** [reply]
Ok thanks @azrikahar , I have tried to add this way using the sdk and it works also well.
```
directus.items<string, Wishes_Users>('wishes_users').updateOne(':wishes_users_Id', {
    products: [ { products_id: ':productsId'}]
}
```

However I'm still stuck with the permissions on Wishes_users_products collection I described upper. 
@joselcvarela @rijkvanzanten does it makes sense I must allow create permissions on the intermediate table to be able to add a wish?

Thanks for help

**Reply by rijkvanzanten** [reply]
> does it makes sense I must allow create permissions on the intermediate table to be able to add a wish?

Yes it does! A many-to-many is no more than a one-to-many to a junction table, which in turn has a many-to-one to another table. Therefore, permissions need to be configured for both the junction table, and the related table (depending on what you want people to be able to do).

**Comment by gkielwasser** [comment]
> Hi, I have the following collection:
> 
> ```
> Wishes_users: {
> 	products: M2M Products,
> 	user: M2O Directus_users
> }
> ```
> 
> I'm trying to restrict using permissions the creation of new items in Wishes_users_products so that it must belongs to the user.
> 
> ```
> POST items/wishes_users_products
> {
> 	"products_id" : "49c67628-eb93-4dc0-89d1-138d249b5c98",
> 	"wishes_users_id":"c64da757-de9a-4105-a897-37cad75a7ccc"
> }
> ```
> 
> Permissions on Wishes_products (Create)
> 
> ```
> {
>     "wishes_users_id": {
>         "user": {
>             "_eq": "$CURRENT_USER"
>         }
>     }
> }
> ```
> 
> As soon as I have this permission defined, I have the following error:
> 
> ```
> {
>   "errors": [
>     {
>       "message": "\"wishes_users_id\" must be of type object",
>       "extensions": {
>         "code": "FAILED_VALIDATION",
>         "field": "wishes_users_id"
>       }
>     }
>   ]
> }
> ```
> 
> Note: about read access of this collection, I have defined the same permission and it works well according to the need describe upper.
> 
> What is wrong with such a permission in the create ?
> 
> Thanks

@rijkvanzanten Could you please help me restricting access on this junction table. I would like to restrict so that it must belongs to the user. I'm not sure what kind of rule I should apply for the create permission. The above rule doesn't work. However it is working well when I set it for the read permission: the user can only see its own wishes.
Thanks

**Reply by rijkvanzanten** [reply]
With these permissions:

```json
{
    "wishes_users_id": {
        "user": {
            "_eq": "$CURRENT_USER"
        }
    }
}
```

you're effectively saying "wishes_users_id has to be an object with a field called user that has to equal to 2f957b12-2246-44e5-9407-d5e806820d01". If you then POST the raw value "2f957b12-2246-44e5-9407-d5e806820d01", it will correctly fail the validation there. 

Try using

```json
{
  "_or": [
    {
      "wishes_users_id": {
        "user": {
          "_eq": "$CURRENT_USER"
        }
      }
    },
    {
      "wishes_users_id": {
        "_eq": "$CURRENT_USER"
      }
    }
  ]
}
```

**Reply by gkielwasser** [reply]
Thanks @rijkvanzanten ! 
Thanks for help @joselcvarela, That was the trick with the "_or" condition.

**Comment by joselcvarela** [comment]
Hey @gkielwasser 
Can you show the columns for the table `wishes_users_products`?

**Reply by gkielwasser** [reply]
Hey @joselcvarela ,
Here are my columns
![wishes_users_products](https://user-images.githubusercontent.com/1532438/136797850-d58d45d1-7687-42cc-a75b-baed2e181d5b.png)
,
Also I have attached the fields I had to autorize in this junction table to be able to add a product wish in wishes_users
![fields](https://user-images.githubusercontent.com/1532438/136798320-cf5a7da4-e289-4350-a446-74baa06f4f93.png)

**Reply by joselcvarela** [reply]
I have tried with `wishes_users` and `wishes_users_products` only and it seems in fact this do not work. Maybe it's because is a nested relationship.
Do you mind opening an issue? Please describe `wishes_users` and `wishes_users_products`.
